### PR TITLE
Bugfix/design flaw in ring buffer

### PR DESCRIPTION
**Pull request by twollgam**
### What does this PR do?

Fixes a design flaw in class RingBuffer. The size cannot reach capacity but this is expected. This PR will fix that.

Additionally RingBuffer is moved to own header file, implementation and declaration is divided and forward declarations for Analyzer is used to prevent to much recompilation while developing.

Add unittests for RingBuffer class.
 
### Closes Issue(s)

None.

### Motivation

Classes should work as expected.

### CODE REVIEW

**Comment by Baklap4** [comment]
I'll take a look at this tonight! Thanks for the PR already :)

File: game/audio.cc [hunk 1]
```diff
@@ -570,8 +580,8 @@ portaudio::Init Audio::init;
 Audio::Audio() {
 	aubio_tempo_set_silence(Audio::aubioTempo.get(), -50.0f);
 	aubio_tempo_set_threshold(Audio::aubioTempo.get(), 0.4f);
-        populateBackends(portaudio::AudioBackends().getBackends());
-        self = std::make_unique<Impl>();
+		populateBackends(portaudio::AudioBackends().getBackends());
+		self = std::make_unique<Impl>();
```
Comment (Baklap4) [inline game/audio.cc:584]: please allign properly

File: game/audio.hh [hunk 2]
```diff
@@ -79,7 +76,7 @@ class ConfigItem;
 /** @short High level audio playback API **/
 class Audio {
 	friend int getBackend();
-        // static because Port audio once for the whole software lifetime
+		// static because Port audio once for the whole software lifetime
```
Comment (Baklap4) [inline game/audio.hh:79]: Align properly

File: game/pitch.hh [hunk 3]
```diff
@@ -117,6 +85,11 @@ public:
 	/** Returns the id (color name) of the mic */
 	std::string const& getId() const { return m_id; }
 
+private:
+	bool calcFFT();
+	void calcTones();
+	void mergeWithOld(tones_t& tones) const;
+
```
Comment (Baklap4) [inline game/pitch.hh:92]: please do not define double private sections
Do keep the methods together and fields together though

I'd prefer the order of method - member

File: game/ringbuffer.hh [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	template <typename InIt> void insert(InIt begin, InIt end);
+	/// Read data from current position if there is enough data to fill the range (otherwise return false). Does not move read pointer.
+	template <typename OutIt> bool read(OutIt begin, OutIt end);
+	void pop(std::size_t n); ///< Move reading pointer forward.
+	std::size_t size() const;
+
+  private:
+	static std::size_t modulo(std::size_t idx);  ///< Modulo operation with proper rounding (handles slightly "negative" idx as well)
+
+  private:
```
Comment (Baklap4) [inline game/ringbuffer.hh:20]: please do not define double private sections
Do keep the methods together and fields together though

I'd prefer the order of method - member

File: game/pitch.hh [hunk 5]
```diff
@@ -31,50 +32,17 @@ static inline bool operator>=(Tone const& lhs, Tone const& rhs) { return lhs.fre
 static inline bool operator<(Tone const& lhs, Tone const& rhs) { return lhs.freq < rhs.freq && lhs != rhs; }
 static inline bool operator>(Tone const& lhs, Tone const& rhs) { return lhs.freq > rhs.freq && lhs != rhs; }
 
-static const unsigned FFT_P = 10;
+static const unsigned FFT_P = 12;
```
Comment (Baklap4) [inline game/pitch.hh:35]: Why did you increase this value?

**Review by Baklap4 — CHANGES_REQUESTED** [review]
Overall looks alright to me. Some minor items to fix as it's mainly refactoring and moving stuff

File: game/pitch.hh (hunk 5 above)
Comment (twollgam) [reply game/pitch.hh:35]: Ups, good catch. Accident :-(

**Comment by ooshlablu** [comment]
Tested with vocals + guitar with the deb and appimage on Ubuntu 20.04. Didn't notice any difference in gameplay, which I think is the point. LGTM

**Review by Baklap4 — APPROVED** [review]
LGTM, discussions have been addressed! Thanks

**Pull request merged**